File: 0087-scramble-string/0087-scramble-string.py
class Solution(object):
    def isScramble(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """  
        n = len(s1)
        dp = [[[False] * (n + 1) for _ in range(n)] for _ in range(n + 1)]
        for k in range(1, n + 1):
            for i in range(n - k + 1):
                for j in range(n - k + 1):
                    if k == 1:
                        dp[i][j][k] = s1[i] == s2[j]
                        continue
                    for u in range(1, k):
                        o1 = dp[i][j][u] and dp[i + u][j + u][k - u]
                        o2 = dp[i][j + k - u][u] and dp[i + u][j][k - u]
                        if o1 or o2:
                            dp[i][j][k] = True
                            break
        return dp[0][0][n]
        

File: 0087-scramble-string/0087-scramble-string.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        int n = s1.length();
        if(s1 == s2) return true;
        vector<vector<vector<bool>>> dp(n, vector<vector<bool>>(n, vector<bool>(n + 1, false)));
        
        for(int k = 1; k <= n; ++k) {
            for(int i = 0; i + k - 1< n; ++i) {
                for(int j = 0; j + k - 1< n; ++j) {
                    if(k == 1) {
                        dp[i][j][k] = s1[i] == s2[j];
                        continue;
                    }
                    
                    for(int u = 1; u < k; ++u) {
                        bool o1 = dp[i][j][u] && dp[i + u][j + u][k - u];
                        bool o2 = dp[i][j + k - u][u] && dp[i + u][j][k - u];
                        if(o1 || o2) {
                            dp[i][j][k] = true;
                            break;
                        }
                    }
                }
            }
        }
        
        return dp[0][0][n];
            
    }
};